### PR TITLE
Do not watch Kotlin caches

### LINKED ISSUE #11511: Make Kotlin DSL caches well-known locations
---
cc: @gradle/build-cache

### PR DESCRIPTION
**Pull request by lptr**
Fixes #11511.

### CODE REVIEW

File: subprojects/snapshots/snapshots.gradle.kts [hunk 1]
```diff
@@ -45,6 +45,11 @@ dependencies {
     testRuntimeOnly(project(":workers"))
     testRuntimeOnly(project(":dependencyManagement"))
 
+    integTestRuntimeOnly(project(":apiMetadata"))
+    integTestRuntimeOnly(project(":kotlinDsl"))
+    integTestRuntimeOnly(project(":kotlinDslProviderPlugins"))
+    integTestRuntimeOnly(project(":kotlinDslToolingBuilders"))
+
```
Comment (eskatos) [inline subprojects/snapshots/snapshots.gradle.kts]: :o: I would move those `integTest*` dependencies below all `test*` declarations

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/provider/KotlinScriptEvaluator.kt [hunk 2]
```diff
@@ -254,7 +258,7 @@ class StandardKotlinScriptEvaluator(
 
         private
         val cacheKeyPrefix =
```
Comment (eskatos) [inline subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/provider/KotlinScriptEvaluator.kt]: :o: I would rename to `cacheKeySpecPrefix` for consistency

File: subprojects/snapshots/snapshots.gradle.kts (hunk 1 above)
Comment (lptr) [reply subprojects/snapshots/snapshots.gradle.kts]: They are underneath all `test*` dependencies, the ones below are for test fixtures. I'd leave it like this unless you feel strongly.

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/provider/KotlinScriptEvaluator.kt (hunk 2 above)
Comment (lptr) [reply subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/provider/KotlinScriptEvaluator.kt]: Renamed.

File: subprojects/snapshots/snapshots.gradle.kts (hunk 1 above)
Comment (eskatos) [reply subprojects/snapshots/snapshots.gradle.kts]: Well, we don't have `integTestFixtures*` yet but we should, so I would move the `integTest*` dependencies below

File: subprojects/core/src/main/java/org/gradle/internal/vfs/AdditiveCacheLocations.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+ * limitations under the License.
+ */
+
+package org.gradle.internal.vfs;
+
+/**
+ * Identifies if a path is underneath one of Gradle's "additive caches".
+ *
+ * Over time more files can be added to an additive cache, but existing files
+ * can never be modified. Files in additive caches can only be deleted as part
```
[the remark this answers is not in the document]
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/internal/vfs/AdditiveCacheLocations.java]: I don't know about such a convention, and I haven't seen it applied consistently. Looks a bit weird, too, especially with the typically short line lenghts we use with Javadoc.

File: subprojects/core/src/main/java/org/gradle/internal/vfs/AdditiveCache.java [hunk 4]
```diff
@@ -14,17 +14,21 @@
...
+ * Represents a store of cached and immutable files.
+ *
+ * Over time more files can be added to an additive cache, but existing files
+ * can never be modified. Files in additive caches can only be deleted as part
+ * of cleanup.
  */
-public interface CachedJarFileStore {
+public interface AdditiveCache {
     /**
      * Returns the root directories of the store.
```
[the remark this answers is not in the document]
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/internal/vfs/AdditiveCache.java]: I think the wording is in line with the Javadoc on the class. How would you change it?

File: subprojects/snapshots/snapshots.gradle.kts (hunk 1 above)
Comment (lptr) [reply subprojects/snapshots/snapshots.gradle.kts]: OK, I interpret this as you feel strongly. I moved them to the bottom.

File: subprojects/core/src/main/java/org/gradle/internal/vfs/AdditiveCache.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/internal/vfs/AdditiveCache.java]: I would have used `Returns the root directories of the additive cache`.

[the remark this answers is not in the document]
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/internal/vfs/AdditiveCache.java]: Changed.

File: subprojects/core/src/main/java/org/gradle/internal/vfs/AdditiveCacheLocations.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+ * of cleanup.
+ *
+ * This quasi-immutability about additive caches allows for some
+ * optimizations wrt retaining file system state in-memory.
```
[the remark this answers is not in the document]
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/internal/vfs/AdditiveCacheLocations.java]: I've touched up the doc a bit.

File: subprojects/core/src/main/java/org/gradle/internal/vfs/RoutingVirtualFileSystem.java [hunk 6]
```diff
@@ -28,13 +27,13 @@
 import java.util.function.Function;
 
 public class RoutingVirtualFileSystem implements VirtualFileSystem {
-    private final WellKnownFileLocations gradleUserHomeFileLocations;
+    private final AdditiveCacheLocations gradleUserHomeFileLocations;
```
[the remark this answers is not in the document]
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/internal/vfs/RoutingVirtualFileSystem.java]: Done

File: subprojects/core/src/test/groovy/org/gradle/internal/vfs/RoutingVirtualFileSystemTest.groovy [hunk 7]
```diff
@@ -16,8 +16,7 @@
 
 package org.gradle.internal.vfs
 
-import org.gradle.api.internal.changedetection.state.DefaultWellKnownFileLocations
-import org.gradle.internal.classpath.CachedJarFileStore
+
```
[the remark this answers is not in the document]
Comment (lptr) [reply subprojects/core/src/test/groovy/org/gradle/internal/vfs/RoutingVirtualFileSystemTest.groovy]: Removed.

File: subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/cache/DefaultGeneralCompileCaches.java [hunk 8]
```diff
@@ -50,7 +50,7 @@
     private final PersistentCache cache;
     private final PersistentIndexedCache<String, PreviousCompilationData> previousCompilationCache;
 
-    public DefaultGeneralCompileCaches(VirtualFileSystem virtualFileSystem, UserHomeScopedCompileCaches userHomeScopedCompileCaches, CacheRepository cacheRepository, Gradle gradle, InMemoryCacheDecoratorFactory inMemoryCacheDecoratorFactory, WellKnownFileLocations fileLocations, StringInterner interner) {
+    public DefaultGeneralCompileCaches(VirtualFileSystem virtualFileSystem, UserHomeScopedCompileCaches userHomeScopedCompileCaches, CacheRepository cacheRepository, Gradle gradle, InMemoryCacheDecoratorFactory inMemoryCacheDecoratorFactory, AdditiveCacheLocations fileLocations, StringInterner interner) {
```
[the remark this answers is not in the document]
Comment (lptr) [reply subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/cache/DefaultGeneralCompileCaches.java]: Done

File: subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/classpath/SplitClasspathEntrySnapshotCache.java [hunk 9]
```diff
@@ -30,19 +30,19 @@
  * All other files are cached in the local cache. Closing this cache only closes the local delegate, not the global one.
  */
 public class SplitClasspathEntrySnapshotCache implements ClasspathEntrySnapshotCache, Closeable {
-    private final WellKnownFileLocations fileLocations;
+    private final AdditiveCacheLocations fileLocations;
```
[the remark this answers is not in the document]
Comment (lptr) [reply subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/classpath/SplitClasspathEntrySnapshotCache.java]: Done

File: subprojects/core/src/main/java/org/gradle/cache/internal/SplitFileContentCacheFactory.java [hunk 10]
```diff
@@ -48,23 +48,23 @@ public void close() throws IOException {
...
         FileContentCache<V> localCache = localFactory.newCache(name, normalizedCacheSize, calculator, serializer);
-        return new SplitFileContentCache<V>(globalCache, localCache, wellKnownFileLocations);
+        return new SplitFileContentCache<V>(globalCache, localCache, additiveCacheLocations);
     }
 
     private static final class SplitFileContentCache<V> implements FileContentCache<V> {
         private final FileContentCache<V> globalCache;
         private final FileContentCache<V> localCache;
-        private final WellKnownFileLocations wellKnownFileLocations;
+        private final AdditiveCacheLocations additiveCacheLocations1;
```
[the remark this answers is not in the document]
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/cache/internal/SplitFileContentCacheFactory.java]: Done

**Pull request merged**